webapp/graphite/worker_pool/pool.py

```diff
@@ -1,58 +1,117 @@
+    self.func = func
+    self.args = args
+    self.kwargs = kwargs
+    self.result = None
+    self.exception = None
+
+  def run(self):
+    try:
+      self.result = self.func(*self.args, **self.kwargs)
+    except Exception as err:
+      self.exception = err
 
 
 def get_pool(name="default", thread_count=settings.POOL_WORKERS):
   """Get (and initialize) a Thread pool.
 
+  If settings.USE_WORKER_POOL is False or thread_count (default: settings.POOL_WORKERS)
+  is 0, then None is returned.
+
   If the thread pool had already been initialized, thread_count will
   be ignored.
   """
   if not settings.USE_WORKER_POOL or not thread_count:
     return None
 
-  with __init_lock:
-    pool = __pools.get(name, None)
+  with _init_lock:
+    pool = _pools.get(name, None)
     if pool is None:
       pool = ThreadPool(thread_count)
-      __pools[name] = pool
+      _pools[name] = pool
   return pool
 
 
 def stop_pools():
-  with __init_lock:
-    for name in list(__pools.keys()):
-      pool = __pools.pop(name)
+  with _init_lock:
+    for name in list(_pools.keys()):
+      pool = _pools.pop(name)
       pool.close()
 
 
 def stop_pool(name="default"):
-  with __init_lock:
-    __pools[name].close()
-    del __pools[name]
+  with _init_lock:
+    _pools[name].close()
+    del _pools[name]
 
 
-def pool_apply(pool, jobs):
-  """Return a queue that will contain results of jobs.
+class PoolTimeoutError(Exception):
+  pass
 
-  A list of jobs is passed in arguments and a queue of
-  results is returned. If settings.USE_WORKER_POOL is
-  True, then the jobs will be applied asynchronously.
+
+def pool_exec(pool, jobs, timeout):
+  """Execute a list of jobs, yielding each one as it completes.
+
+  If a pool is specified then the jobs will be executed asynchronously,
+  otherwise they are executed in order.
+
+  If not all jobs have been executed after the specified timeout a
+  PoolTimeoutError will be raised. When operating synchronously the
+  timeout is checked after each job is run.
   """
-  queue = Queue.Queue()
+  start = time.time()
+  deadline = start + timeout
   if pool:
-    def return_result(x):
-      return queue.put(x)
+    queue = Queue.Queue()
+
+    def pool_executor(job):
+      job.run()
+      queue.put(job)
+
     for job in jobs:
-      pool.apply_async(
-        func=job[0], args=job[1:], callback=return_result)
+      pool.apply_async(func=pool_executor, args=[job])
+
+    done = 0
+    total = len(jobs)
+
+    while done < total:
+      wait_time = max(0, deadline - time.time())
+      try:
+        job = queue.get(True, wait_time)
+      except Queue.Empty:
+        raise PoolTimeoutError("Timed out after %fs" % (time.time() - start))
+
+      done += 1
+      yield job
   else:
     for job in jobs:
-      queue.put(job[0](*job[1:]))
-  return queue
+      job.run()
+
+      if time.time() > deadline:
+        raise PoolTimeoutError("Timed out after %fs" % (time.time() - start))
+
+      yield job
```

Comment:
`pool_exec` takes care of a few things that were previously duplicated in different places that used `pool_apply`.  It takes care of tracking timeouts (and adds timeout checking between jobs if the worker pool isn't used), it also allows a caller to just iterate over the result instead of having to deal with the queue directly.